torchao/prototype/quantized_training/int8_mixed_precision.py

```diff
@@ -0,0 +1,222 @@
+    def __new__(cls, data: Tensor, config: Int8MixedPrecisionTrainingConfig):
+        return Tensor._make_wrapper_subclass(
+            cls,
+            data.shape,
+            data.stride(),
+            data.storage_offset(),
+            dtype=data.dtype,
+            device=data.device,
+        )
+
+    @torch._dynamo.disable
+    def __init__(self, data: Tensor, config: Int8MixedPrecisionTrainingConfig):
+        self._data = data
+        self.config = config
+
+    def __tensor_flatten__(self):
+        return ["_data"], [self.config]
+
+    @classmethod
+    def __tensor_unflatten__(cls, tensor_data_dict, tensor_attributes, outer_size=None, outer_stride=None):
+        return cls(tensor_data_dict["_data"], *tensor_attributes)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(data={self._data}, config={self.config})"
+
+    def to_original(self):
+        return self._data.clone()
+
+    # adapated from FP8 implementation of WeightWithDynamicFloat8CastTensor
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        config = None
+
+        def unwrap(x: cls):
+            nonlocal config
+            if config is None:
+                config = x.config
+            else:
+                assert x.config == config
+            return x._data
+
+        out = func(
+            *pytree.tree_map_only(cls, unwrap, args),
+            **pytree.tree_map_only(cls, unwrap, kwargs),
+        )
+
+        if func is aten.copy_.default:
+            # return original object
+            return args[0]
+        elif func in {
+            aten.t.default,
+            aten.detach.default,
+            aten.empty_like.default,
+            aten.new_zeros.default,
+            aten.slice.Tensor,
+            aten.view.default,
+            aten.as_strided.default,
+            aten._to_copy.default,
+            aten._pin_memory.default,
+            aten.split.Tensor,
+            aten.clone.default,
+        }:
+            # return new wrapped object
+            return pytree.tree_map_only(Tensor, lambda x: cls(x, config), out)
+        else:
+            # return new unwrapped object
+            return out
+
+    def fsdp_pre_all_gather(self, mesh):
+        # TODO: pre-quantize weight here -> reduce comm bandwidth.
+        # we will need another tensor subclass to hold the quantized weight.
+        return (self._data,), (self.config,)
+
+    def fsdp_post_all_gather(
+        self,
+        all_gather_outputs: Tuple[Tensor, ...],
+        metadata: Any,
+        param_dtype: torch.dtype,
+        *,
+        out: Optional[Tensor] = None,
+    ):
+        (data,) = all_gather_outputs
+        (config,) = metadata
+        if out is not None:
+            assert isinstance(out, Int8MixedPrecisionTrainingLinearWeight)
+            assert out.config == config
+            return
+        return Int8MixedPrecisionTrainingLinearWeight(data.to(param_dtype), config), all_gather_outputs
```

Comment:
@vkuzo The only thing not working with FSDP now (in both eager and compile mode) is `MixedPrecisionPolicy(param_dtype=torch.bfloat16)` with FP32 model. I think this is where it goes wrong. With this code, the loss diverges from baseline.

If I change to the below, loss becomes NaN.

```python
data = data.to(param_dtype)
return Int8MixedPrecisionTrainingLinearWeight(data, config), (data,)
```

If I don't do anything about `param_dtype`, I will get `AssertionError: FSDP reduce-scatter expects uniform gradient dtype but got {torch.bfloat16, torch.float32}`. 

Do you know what should be the correct way to do this? Also, what should `fsdp_post_all_gather()` return? I suppose this is a private API so there is no documentation.

I think one way to circumvent this is to override `dtype` of tensor subclass and keeps track of this metadata to propagate across ops (like NF4), and in matmul/linear, return the correct dtype. However, it seems a bit cumbersome for what seems to be simple.

Reply:
Not directly answering your question, but here is how I built / encouraged folks to build things for float8 training:
1. focus on promising accuracy + performance first, take the simplest possible path through the PT stack to get good numbers
2. improve UX, as needed, after (1) is done

At a high level, that meant using module swaps and sidestepping a lot of missing functionality in torch.compile + tensor subclasses, to move faster on (1).  If you want, you are welcome to do the same here.  That would mean something like:
1. have a Int8TraningLinear, weight is a regular tensor
2. tensor subclasses are transient and only used to work around autograd limitations and compose with distributed
3. if you follow what float8 is doing, things will work with compile/distributed/autograd as we've fixed most/all of the issues

Note that this isn't aligned with torchao inference, and that alignment will have to happen at some future time, I'm just decoupling that from unblocking early research / getting promising performance + accuracy.

Reply:
> Also, what should fsdp_post_all_gather() return?

This is an API designed to allow FSDP to communicate in lower precision.  Since you are currently always wrapping weights with subclasses, something that would sound right to me is:

* first, just make FSDP comms work in higher precision, this means you have to ser/deser your weight wrapper in these APIs
* then, in a separate PR, optinally enable doing the comms in int8, this means you quantize in the fsdp_pre_all_gather API and reconstruct the already quantized tensor in the fsdp_post_all_gather API

Reply:
> I think one way to circumvent this is to override dtype of tensor subclass and keeps track of this metadata to propagate across ops (like NF4), and in matmul/linear, return the correct dtype. However, it seems a bit cumbersome for what seems to be simple.

hmm, not sure exactly what this means, but for float8, dtype is set to what we want autograd to see.  I understand this might be confusing to a subset of people, but I think being practical and just having things work e2e with a slight regression to what is intuitive has worked well in that case.  There isn't a single thing we can use as a "dtype" for arbitrary tensor subclasses anyways which composes with all the PyTorch systems.

Reply:
Yea I think "module swap + regular weight tensor + call custom autograd.Function in forward" is simpler to implement too. May need to add `@torch._dynamo.allow_in_graph` to the custom autograd.Function for compile to work. I tested this combo and it works for single GPU. Will try if it works for FSDP w/ mixed-precision param_dtype.

The main reason I started this out as tensor subclass is because of familiarity when I wrote other stuff in torchao. Also, for quantized weight before all-gather, tensor subclass for linear weight is needed (similar to `WeightWithDynamicFloat8CastTensor`) (though I agree that this shouldn't be a big reason at this stage). Anyway, single GPU performance is already pretty good 😄, only FSDP is left. Though I guess FSDP performance is more important for big corps 😄.

> This is an API designed to allow FSDP to communicate in lower precision

Yea I realize if the tensor subclass doesn't implement this method, the following ops will be called

```python
torch.ops.fsdp.all_gather_copy_in.default,
torch.ops.fsdp.split_with_sizes_copy.default,
torch.ops.c10d._allgather_base_.default,
```

It's not clear which op here should return the wrapped tensor (so that later `F.linear()` dispatch is correct). So maybe it's also possible to not use `fsdp_pre/post_all_gather()` if I don't need low-precision comms at the moment.

> first, just make FSDP comms work in higher precision, this means you have to ser/deser your weight wrapper in these APIs

Yes, this is my intention. But right now it doesn't work with `MixedPrecisionPolicy(param_dtype=torch.bfloat16)` as I said previously. There will be complaints about dtype (thus I thought about keeping tracking of a separate dtype metadata like NF4). If there is no MixedPrecisionPolicy, the current state of this PR works with FSDP, but very slow even with torch.compile.

Reply:
> It's not clear which op here should return the wrapped tensor (so that later F.linear() dispatch is correct).

without taking a close look at the code, I think a good answer would be to handle all of these ops in your subclass properly.  Otherwise, I would expect the logic to get pretty brittle / hard to debug real fast.

Reply:
Just tested this combo "module swap + regular weight tensor + call custom autograd.Function in forward + `@torch._dynamo.allow_in_graph`" and it works with FSDP + MixedPrecisionPolicy now 😭. Seeing +15% speedup with 8x A100 for Llama3-8B pre-training. (for 1x A100, I see +40% speedup for Llama2-7B pre-training).

For `fsdp_post_all_gather()`, I guess I'm not supposed to change the data here, including dtype casting. Argument `param_dtype` should only be used to create "appearance" dtype for the tensor subclass. Maybe you can help me confirm it with FSDP people?

I will add back the module swap UX. I think I can leave the tensor subclass as is for future PR to implement INT8 comms (though right now I'm seeing a problem in backward all-gather 🌚 - I'm doing row-wise scaling for weight in forward, which is fine since weight is sharded along dim0, but in backward, I'm doing col-wise scaling (since we need to compute `grad_weight = grad_output.T @ input`) -> not compatible with dim0 sharding. Perhaps I need to support tensor-wise scaling for weight, which is the same as [SwitchBack](https://arxiv.org/abs/2304.13013) and FP8 recipes, though I suspect tensor-wise scaling for INT8 will have problems... OR tile-wise scaling (which requires more customized triton kernel...)

Reply:
for some additional context, my high level plan for float8 was
1. eventually remove the module swap UX in favor of either module hooks or torch_function overrides (like you have in this PR), for better handling of unknown unknowns.  This will likely involve a good amount of e2e torch.compile work to ensure the new solution does not regress accuracy and performance
2. align with torchao inference, since (1) will not be doing exactly what torchao inference is doing since we will need either hooks or torch_function overrides, due to training-specific functionality.

Neither of these have a timeline yet, I think 2024Q4 would be a good time to start on (1), and (2) will depend on specifics of (1).  cc @jerryzh168 , @msaroufim  for context.

Reply:
@vkuzo I got FSDP w/ MixedPrecisionPolicy working for tensor subclass UX too! The trick is to do dtype casting in `fsdp_pre_all_gather()` instead of `fsdp_post_all_gather()`. Sadly `param_dtype` is not exposed to pre all-gather, so I think I need to add an extra field in `Int8MixedPrecisionTrainingConfig` as a workaround.

The performance is better than module swap API too! I'm seeing 20% speedup with tensor subclass compared to 15% speedup with module swap. (I think it might have to do with the use of `@torch._dynamo.allow_in_graph`, and since I put my quantization code inside the custom `autograd.Function`, it will not compile the quantization code?)

Reply:
> I got FSDP w/ MixedPrecisionPolicy working for tensor subclass UX too!

that's great.  The torch_function+compile path was not ready in the past when float8 was built, great to hear that the story is better now.  If everything in fact works e2e nowadays with this approach, I could see float8 migrating to something like this in the future to unify with the rest of torchao better.

> The performance is better than module swap API too! I'm seeing 20% speedup with tensor subclass compared to 15% speedup with module swap.

that sounds like a bug, lmk if you have logs from a small toy model to demonstrate what is different, I'd be interested to check that out

Reply:
> that sounds like a bug, lmk if you have logs from a small toy model to demonstrate what is different, I'd be interested to check that out

What kind of logs do you need? Like profile trace? I don't keep this code anymore, so need some efforts to reproduce it. If you really want to check it out, I can produce it for you.

Also a side observation. Without `fsdp_pre/post_all_gather()`, I think we can't propagate tensor subclass across FSDP all-gather. I tried wrapping the outputs of the 3 FSDP ops mentioned earlier (`fsdp.all_gather_copy_in.default`, ...) but still could not propagate the tensor subclass. I think this is why NF4 also needs to use `fsdp_pre/post_all_gather()`, even though it doesn't exactly do dist comm quantization (just all-gather the pre-quantized weight, and wrap it with tensor subclass again).

Reply:
> What kind of logs do you need? Like profile trace? I don't keep this code anymore, so need some efforts to reproduce it. If you really want to check it out, I can produce it for you.

feel free to ignore since looks like it will be a lot of work.  I can check it out from my machine after this lands, although it will be on different hardware.

Reply:
btw, just curious, do you have end2end speedup numbers for FP8? Specifically for Llama3-8B. I see there is one here https://github.com/pytorch/torchtitan/blob/main/docs/performance.md, but it's with tensor parallel and for Llama3-405B.

Reply:
this is the most recent polished numbers we have: https://dev-discuss.pytorch.org/t/enabling-float8-all-gather-in-fsdp2/2359 .  It's on LLaMa 3 70B, 1.5x speedup on 128 H100s with FSDP but no TP/SP.

Reply:
The setup there seems like doesn't use many special tricks yet (still just dynamic scaling) + FP8 all-gather (which probably is not too significant for an 8x A100 setup like I used). So there are still room to optimize INT8 training further :smile: (assuming INT8 and FP8 speedups over BF16 are similar) (or maybe using Llama3-70B will have larger speedup simply thanks to larger matmul -> more % time spent in matmul)